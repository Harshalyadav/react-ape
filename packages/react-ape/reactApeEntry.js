/**
 * Copyright (c) 2018-present, Raphael Amorim.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import ReactApeRenderer from './renderer/ReactApeRenderer';
import StyleSheetModule from './modules/StyleSheet';

export const render = ReactApeRenderer.render;
// export const unmountComponentAtNode = ReactTVRenderer.unmountComponentAtNode;

export const Canvas = 'CANVAS';
export const ListView = 'LISTVIEW';
export const Image = 'IMAGE';
export const View = 'VIEW';
export const Text = 'TEXT';
export const StyleSheet = StyleSheetModule;

export default ReactApeRenderer;
