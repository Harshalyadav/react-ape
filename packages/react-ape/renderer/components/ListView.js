class ListView {
  constructor(props) {
    this.props = props;

    this._renderList = [];
  }

  appendChild(fn) {
    this._renderList.push(fn);
  }

  children() {
    return this._renderList;
  }

  render(apeContext) {
    const { ctx } = apeContext;
    const { style = {}, dataSource, renderRow } = this.props;

    const callRenderFunctions = (renderFunction) => {
      renderFunction(apeContext);
    }

    this._renderList.forEach(callRenderFunctions);

    console.log(dataSource, renderRow)

  }
}

export default ListView;
