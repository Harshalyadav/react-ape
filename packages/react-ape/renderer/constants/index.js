/**
 * Copyright (c) 2018-present, Raphael Amorim.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

// Defaults for Render
export const ViewDefaults = {
  size: 200, // 200x200
  lineHeight: 24,
};

// ReactApe Internal Constants
export const _SectionBlockSize: number = 80; // 80x80
