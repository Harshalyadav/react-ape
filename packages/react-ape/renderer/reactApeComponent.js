import Image from './components/Image';
import ListView from './components/ListView';
import Text from './components/Text';
import View from './components/View';

const CHILDREN = 'children';
const STYLE = 'style';

const ReactApeComponent = {
 createElement(type, props, rootContainerElement, internalInstanceHandle) {

    const COMPONENTS = {
      IMAGE: () => Image.bind(this, props),
      TEXT: () => Text.bind(this, props),

      LISTVIEW: () => new ListView(props),
      VIEW: () => new View(props),
    };

    return (
      COMPONENTS[type] ?
      COMPONENTS[type]() :
      console.warn('ReactApe could not identify ' + type + ' as ReactApeComponent')
    )
  },

  createTextNode(text, rootContainerElement) {
    return text;
    // return getOwnerDocumentFromRootContainer(
    //   rootContainerElement
    // ).createTextNode(text);
  },

  diffProperties(element, tag, lastRawProps, nextRawProps, rootContainerElement) {
    let updatePayload = null;

    const lastProps = lastRawProps;
    const nextProps = nextRawProps;

    let propKey;
    let styleName;
    let styleUpdates = null;
    for (propKey in lastProps) {
      if (
        nextProps.hasOwnProperty(propKey) ||
        !lastProps.hasOwnProperty(propKey) ||
        lastProps[propKey] == null
      ) {
        continue;
      }
      if (propKey === STYLE) {
        const lastStyle = lastProps[propKey];
        for (styleName in lastStyle) {
          if (lastStyle.hasOwnProperty(styleName)) {
            if (!styleUpdates) {
              styleUpdates = {};
            }
            styleUpdates[styleName] = '';
          }
        }
      } else {
        // For all other deleted properties we add it to the queue. We use
        // the whitelist in the commit phase instead.
        (updatePayload = updatePayload || []).push(propKey, null);
      }
    }
    for (propKey in nextProps) {
      const nextProp = nextProps[propKey];
      const lastProp = lastProps != null ? lastProps[propKey] : undefined;
      if (
        !nextProps.hasOwnProperty(propKey) ||
        nextProp === lastProp ||
        (nextProp == null && lastProp == null)
      ) {
        continue;
      }
      if (propKey === STYLE) {
        if (lastProp) {
          // Unset styles on `lastProp` but not on `nextProp`.
          for (styleName in lastProp) {
            if (
              lastProp.hasOwnProperty(styleName) &&
              (!nextProp || !nextProp.hasOwnProperty(styleName))
            ) {
              if (!styleUpdates) {
                styleUpdates = {};
              }
              styleUpdates[styleName] = '';
            }
          }
          // Update styles that changed since `lastProp`.
          for (styleName in nextProp) {
            if (
              nextProp.hasOwnProperty(styleName) &&
              lastProp[styleName] !== nextProp[styleName]
            ) {
              if (!styleUpdates) {
                styleUpdates = {};
              }
              styleUpdates[styleName] = nextProp[styleName];
            }
          }
        } else {
          // Relies on `updateStylesByID` not mutating `styleUpdates`.
          if (!styleUpdates) {
            if (!updatePayload) {
              updatePayload = [];
            }
            updatePayload.push(propKey, styleUpdates);
          }
          styleUpdates = nextProp;
        }
      } else if (propKey === CHILDREN) {
        if (
          lastProp !== nextProp &&
          (typeof nextProp === 'string' || typeof nextProp === 'number')
        ) {
          (updatePayload = updatePayload || []).push(propKey, nextProp);
        }
      } else {
        // For any other property we always add it to the queue and then we
        // filter it out using the whitelist during the commit.
        (updatePayload = updatePayload || []).push(propKey, nextProp);
      }
    }
    if (styleUpdates) {
      (updatePayload = updatePayload || []).push(STYLE, styleUpdates);
    }

    return updatePayload;
  }
}

export default ReactApeComponent;
