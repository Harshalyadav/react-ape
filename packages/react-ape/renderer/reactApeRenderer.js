import reconciler from 'react-reconciler';
import reactApeComponent from './reactApeComponent';
import {precacheFiberNode, updateFiberProps} from './reactApeComponentTree';

function scaleDPI(canvas, context, customWidth, customHeight) {
  const width = customWidth ||
              canvas.offsetWidth ||
              canvas.width ||
              canvas.clientWidth;
  const height = customHeight ||
               canvas.offsetHeight ||
               canvas.height ||
               canvas.clientHeight;
  const deviceRatio = window.devicePixelRatio || 1;
  const bsRatio = context.webkitBackingStorePixelRatio ||
                context.mozBackingStorePixelRatio ||
                context.msBackingStorePixelRatio ||
                context.oBackingStorePixelRatio ||
                context.backingStorePixelRatio || 1;
  const ratio = deviceRatio / bsRatio;

  if (deviceRatio !== bsRatio) {
    canvas.width = Math.round(width * ratio);
    canvas.height = Math.round(height * ratio);
    canvas.style.width = width + 'px';
    canvas.style.height = height + 'px';
    context.scale(ratio, ratio);
  }
  return ratio;
};

// TODO: Use Context.
let apeContextGlobal = false;

const ReactApeFiber = reconciler({
  appendInitialChild(parentInstance, child) {
    console.log('appendInitialChild', parentInstance, child);
    if (parentInstance.appendChild) {
      parentInstance.appendChild(child);
      parentInstance.render(apeContextGlobal);
    }
  },

  createInstance(type, props, rootContainerInstance, hostContext, internalInstanceHandle) {
    let apeContext = null;
    if (!apeContextGlobal && rootContainerInstance.getContext) {
      let rootContainerInstanceContext = rootContainerInstance.getContext('2d');

      // TODO: Change it.
      scaleDPI(rootContainerInstance, rootContainerInstanceContext);
      apeContextGlobal = {
        type: 'canvas',
        ctx: rootContainerInstanceContext = rootContainerInstance.getContext('2d'),
      };
    }

    const apeElement = reactApeComponent.createElement(
      type,
      props,
      rootContainerInstance,
      apeContextGlobal,
      internalInstanceHandle,
    );

    // console.log(apeElement, internalInstanceHandle)

    precacheFiberNode(internalInstanceHandle, apeElement);
    updateFiberProps(apeElement, props);
    return apeElement;
  },

  createTextInstance(text, rootContainerInstance, internalInstanceHandle) {
    return text;
  },

  finalizeInitialChildren(element, type, props) {
    return false;
  },

  getPublicInstance(inst) {
    return inst;
  },

  prepareForCommit() {
    // noop
  },

  prepareUpdate(element, type, oldProps, newProps) {
    // console.log(element)
    // TODO: Updates
    return true;
  },

  resetAfterCommit() {
    // noop
  },

  resetTextContent(element) {
    // noop
  },

  getRootHostContext(rootInstance) {
    // You can use this 'rootInstance' to pass data from the roots.
    return rootInstance;
  },

  getChildHostContext() {
    return {};
  },

  shouldSetTextContent(type, props) {
    return false;
  },

  scheduleAnimationCallback() {},

  scheduleDeferredCallback() {},

  useSyncScheduling: true,

  now: Date.now,

  mutation: {
    appendChild(parentInstance, child) {
      console.log('appendChild', parentInstance, child);
      // if (parentInstance.appendChild) {
      //   parentInstance.appendChild(child);
      // } else {
      //   parentInstance.document = child;
      // }
    },

    appendChildToContainer(parentInstance, child) {
      console.log('appendChildToContainer', parentInstance, child)
      if (child.render) {
        child.render(apeContextGlobal);
      } else {
        child(apeContextGlobal);
      }
    },

    removeChild(parentInstance, child) {
      // parentInstance.removeChild(child);
    },

    removeChildFromContainer(parentInstance, child) {
      // parentInstance.removeChild(child);
    },

    insertBefore(parentInstance, child, beforeChild) {
      // noob
    },

    commitUpdate(instance, updatePayload, type, oldProps, newProps) {
      // noop
    },

    commitMount(instance, updatePayload, type, oldProps, newProps) {
      // noop
    },

    commitTextUpdate(textInstance, oldText, newText) {
      // textInstance.children = newText;
    },
  }
})

const defaultContainer = {};
const roots = new Map();

const ReactApeRenderer = {
  render(canvasElement, container, callback) {
    const containerKey =
      typeof container === 'undefined' ? defaultContainer : container;
    let root = roots.get(containerKey);
    if (!root) {
      root = ReactApeFiber.createContainer(containerKey);
      roots.set(container, root);
    }

    ReactApeFiber.updateContainer(canvasElement, root, null, callback);

    ReactApeFiber.injectIntoDevTools({
      bundleType: 1,
      rendererPackageName: 'ReactApe',
      findHostInstanceByFiber: ReactApeFiber.findHostInstance,
    });

    return ReactApeFiber.getPublicRootInstance(root);
  },
}

export default ReactApeRenderer;
