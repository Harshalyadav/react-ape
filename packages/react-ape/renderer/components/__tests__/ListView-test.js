import React from 'react';
import renderer from 'react-test-renderer';

import ListView from '../ListView';
import {View, Text} from '../../../reactApeEntry';

global.CanvasRenderingContext2D = () => {};

describe('ListView', () => {
  it("should render empty view when doesn't exist dataSource", () => {
    const ListViewTree = renderer.create(<ListView />).toJSON();
    expect(ListViewTree).toEqual(<View />);
  });

  it('should render properly with dataSource and renderRow', () => {
    const dataSource = [
      {dog: 'Pug', age: 5},
      {dog: 'Golden Retriever', age: 8},
    ];
    const renderRow = data => (
      <Text>
        {data.dog}, which age is {data.age}
      </Text>
    );

    const ListViewTree = renderer
      .create(<ListView renderRow={renderRow} dataSource={dataSource} />)
      .toJSON();
    expect(ListViewTree).toEqual(
      <View>
        <Text>Pug, which age is 5</Text>
        <Text>Golden Retriever, which age is 8</Text>
      </View>
    );
  });

  it('renders correctly', () => {
    const dataSource = [{name: 'Jack'}, {name: 'Russel'}];
    const renderRow = (data, idx) => (
      <Text id={'render-row-' + idx}>{data.name}</Text>
    );

    const ListViewTree = renderer
      .create(<ListView renderRow={renderRow} dataSource={dataSource} />)
      .toJSON();

    expect(ListViewTree).toMatchSnapshot();
  });
});
