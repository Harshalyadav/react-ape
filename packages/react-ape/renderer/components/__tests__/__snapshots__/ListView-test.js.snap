// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListView renders correctly 1`] = `
<View>
  <Text
    id="render-row-0"
  >
    Jack
  </Text>
  <Text
    id="render-row-1"
  >
    Russel
  </Text>
</View>
`;
